### PR TITLE
feature: support propagation.never, propagation.mandatory, transaction suspend and resume api

### PR DESCRIPTION
**Pull request by lightClouds917**
### Ⅰ. Describe what this PR did
1.add propagation about:never,mandatory；
2.optimize the transationTemplate code；

### CODE REVIEW

**Comment by lightClouds917** [comment]
@CharmingRabbit @zjinlei PTAL

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 1]
```diff
@@ -103,13 +114,43 @@ public Object execute(TransactionalExecutor business) throws Throwable {
...
-            if (previousXid != null) {
-                RootContext.bind(previousXid);
-            }
+            resume(suspendedXid);
         }
 
     }
 
+    private boolean existingTransaction() {
+        return StringUtils.isEmpty(RootContext.getXID()) ? false : true;
```
Comment (zjinlei) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: StringUtils.isEmpty(RootContext.getXID()) has returned boolean, can use notEmpty

Comment (zjinlei) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: StringUtils.isEmpty(RootContext.getXID()) has returned boolean, can use notEmpty

**Comment by CharmingRabbit** [comment]
> @CharmingRabbit @zjinlei PTAL

LGTM.
Thanks for your improvement

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java (hunk 1 above)
Comment (lightClouds917) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: done

Comment (lightClouds917) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: done

**Comment by lightClouds917** [comment]
I created  a new class SuspendedResourcesHolder,Can be used to temporarily save transaction state for later recovery,when we suspend global transaction ,or change the branchType, Or handle other nesting situations later.

**Comment by lightClouds917** [comment]
@zjinlei @CharmingRabbit @a364176773 PTAL,THS

**Comment by lightClouds917** [comment]
@zjinlei  @CharmingRabbit @a364176773 PTAL,THS

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 2]
```diff
@@ -52,29 +53,42 @@ public Object execute(TransactionalExecutor business) throws Throwable {
...
-                    if (StringUtils.isEmpty(RootContext.getXID())) {
+                    if (!existingTransaction()) {
                         return business.execute();
                     }
                     break;
                 case REQUIRED:
                     break;
+                case NEVER:
+                    if (existingTransaction()) {
+                        throw new TransactionException("Existing transaction found for transaction marked with propagation 'never'");
```
Comment (slievrly) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: how about print xid here？

Comment (lightClouds917) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: ok，i will add now

Comment (lightClouds917) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: finished

**Pull request merged**